### DISCUSSION TITLE
Erroneous convergence of Exact sampler for small Hilbert sizes

### DISCUSSION THREAD

**Discussion by Gauthameshwar** [netket/netket#1379] [General]
I have problems with the convergence of the ground state energy of a neural network running Monte Carlo from exact sampling. For simplicity, I have considered a system that is not too large in size (hi.n_size ~ 6300). However, even with exact sampling, my model gives a lot of errors and uncertainty while converging to a value. I tried to closely follow the tutorials in Bose-Hubbard systems [here](https://netket.readthedocs.io/en/latest/tutorials/gs-matrix-models.html) (even made the learning rate small and increased the iterations), but I still face lots of errors in the results! Can anyone please tell me where the issue is coming from? I'm attaching the convergence plot and the code. 

```python
import os
os.environ["JAX_PLATFORM_NAME"] = "cpu"

# Import netket library
import netket as nk
print('Version of Netket: ',nk.__version__)

# Import Json, this will be needed to load log files
import json

# Helper libraries
import numpy as np
import matplotlib.pyplot as plt
import time
import flax.linen
import jax
import jax.numpy as jnp

Lambda = 5  # cutoff of each bosonic Fock space
N = 7  # number of bosons in our system 
L = 9 # no of sites in our bosonic system
g = nk.graph.Hypercube(length=L, n_dim=1, pbc=True)

hi = nk.hilbert.Fock(n_max = Lambda, N = L, n_particles = N)  # n_max -> Maximum occupation for a site

#the system parameters
V = 1.0
J = 1.0
U = 10.0

from netket.operator.boson import create, destroy, number

# we loop over all "sites" -> loop over N bosons PBC is assumed
H_hop = sum([-J * (create(hi, i%L) * destroy(hi, (i+1)%L) + create(hi, (i+1)%L) * destroy(hi, i%L)) for i in range(L)])
    
H_onsite = sum([V * (i - L//2)**2 * (create(hi, i) * destroy(hi, i)) for i in range(L)])
    
H_interact = sum([U/2 * (number(hi, i)*number(hi, i) - number(hi, i)) for i in range(L)])
    
H = H_hop + H_onsite + H_interact
    
#H = nk.operator.BoseHubbard(hi, U = U, J = J, V = V, mu = 0.0, graph = g)
    
eig_vals = nk.exact.lanczos_ed(
      H, compute_eigenvectors=False
    )
    
print('On-site interaction U: %s'%U)
    
print(eig_vals[0])
    
# Creating a neural network for training our system
    
     
g = nk.graph.Hypercube(length=L, n_dim=1, pbc=True)
# hi = nk.hilbert.Fock(n_max=Lambda,n_particles=N, N=g.n_nodes)
    
class Model(nk.nn.Module):
    @nk.nn.compact
    def __call__(self, x):
         
        y = nk.nn.Dense(features=20, dtype=np.complex128, kernel_init=jax.nn.initializers.normal(stddev=0.1), bias_init=jax.nn.initializers.normal(stddev=0.1))(x)
        #y = nk.nn.Dense(features=20, dtype=np.complex128, kernel_init=jax.nn.initializers.normal(stddev=0.1), bias_init=jax.nn.initializers.normal(stddev=0.1))(x)
    
        y = nk.nn.Dense(features=2, dtype=np.complex128, kernel_init=jax.nn.initializers.normal(stddev=0.1), bias_init=jax.nn.initializers.normal(stddev=0.1))(y)
    
        y = nk.nn.activation.tanh(y)
        return jax.numpy.sum(y, axis=-1)
        
ffnn = Model()
sa = nk.sampler.ExactSampler(hi)
    
# The variational state
vs = nk.vqs.MCState(sa, ffnn, n_samples = 1024)
    
op = nk.optimizer.Sgd(learning_rate=0.0005)
    
# Stochastic Reconfiguration
sr = nk.optimizer.SR(diag_shift=.8)
    
# The ground-state optimization loop
gs = nk.VMC(
        hamiltonian=H,
        optimizer=op,
        preconditioner=sr,
        variational_state=vs)
        
iter = 10000
    
start = time.time()
gs.run(out='FF%s'%U, n_iter=iter)
end = time.time()
    
print('### Feed Forward calculation')
print('Avg. E_g: ', vs.expect(H))
    
#model_E.append(vs.expect(H))
    
# import the data from log file
    
data_ff=json.load(open("FF%s.log"%U))
    
iters_ff = data_ff["Energy"]["iters"]
energy_ff = data_ff["Energy"]["Mean"]["real"]

plt.figure()
plt.plot(iters_ff, energy_ff, color='blue', marker=None, linestyle="-",label='Feed_Forward_NN for U = %s'%U)
plt.plot([i for i in [0,iter]],[eig_vals[0] for i in [0,iter]],'--k',label = 'Exact $E_g$ using Lanczos')
plt.legend(loc='right')
plt.ylabel('$E_g$')
plt.xlabel('Iteration')

plt.show()
name = 'Convergence_U_%s.png'%U
plt.savefig(name)

---------------------------------------------------------------------------------------------------------------------

Version of Netket:  3.5.1
On-site interaction U: 10.0
23.88088649911016
C:\Users\gauta\anaconda3\lib\site-packages\netket\nn\__init__.py:91: FutureWarning: 

          =======================================================================
          `nk.nn.Module` is deprecated. Use `flax.linen.Module` directly instead.
          =======================================================================

If you imported `flax.linen as nn`, as is customary to do, you can use `nn.Module` directly. There are no functionality changes.

  _warn_deprecation(
C:\Users\gauta\anaconda3\lib\site-packages\netket\nn\__init__.py:91: FutureWarning: 

          =======================================================================
          `nk.nn.compact` is deprecated. Use `flax.linen.compact` directly instead.
          =======================================================================

If you imported `flax.linen as nn`, as is customary to do, you can use `nn.compact` directly. There are no functionality changes.

  _warn_deprecation(
100%|███████████████| 10000/10000 [57:49<00:00,  2.88it/s, Energy=29.00001808-0.00000701j ± 0.00000020 [σ²=0.00000000]]
### Feed Forward calculation
Avg. E_g:  29.00001809-0.00000721j ± 0.00000021 [σ²=0.00000000]
```

Plot: 

<img width="468" alt="image" src="https://user-images.githubusercontent.com/42206163/199198015-f3237fa4-f75c-4bdd-be14-6ac6c19fff02.png">

**Comment by PhilipVinc** [comment]
@zakari-denis might have some comments because he's working with Bose Hubbard right now but I can tell you that:

 - $U=10$ is a hard regime. I'm not aware of anyone getting good results on large systems. It surfaces several issues at various levels...
 - You should check that your sampling is converged and that you get enough variance to have a signal to optimise
 - Lambda=5 means you have an high cutoff. Using just 20 hidden features for such large Hilbert space is not much. I'd be surprised if it works. 
 - For BH systems, a thing you might try is to use a one-hot encoding before feeding to dense layers. Also, increase the feature density.
 - SR diagonal shift degrades the solution you find but stabilises the optimisation. It's hard to give precise quantities, but 0.8 is very large. I wouldn't be surprised if this also is degrading your result
 - The spikes you see are either because of sampling or because of too large learning rate. 

Posible way out: change your model. Try to start from smaller U and increase it slowly. Don't start immediately with hard simulations.

**Reply by Gauthameshwar** [reply]
Thank you for your comments!

I went through each of them step by step and saw if I was getting better results. Here is what I found:

*   $U = 10$ is the on-site interaction of the model. It only does as much as changing the values of the diagonal terms in the Hamiltonian since $\frac{U}{2} \hat{n}_i (\hat{n}_i -1 )$ is a local interaction in the position space. It doesn't increase the Hilbert size... Will learning it still be a problem if the matrix has large values in the diagonals?
*    I now understand the purpose of SR diagonal shift better. I had a slight intuition since increasing it reduced the fluctuations... But I can't get a balance of convergence and accurate results no matter what I try... Here's what happened when I put shift = 0.1

<img width="319" alt="image" src="https://user-images.githubusercontent.com/42206163/199272213-33569138-c683-4f77-88c3-a2838fd79bce.png">

*   I tried adding more layers, but it didn't help much (the above plot had twice the layers as before). The size of the space is: 
```
hi.n_states
> 6354
```
I thought it was not too big to train since the tutorial BHM had a similar hi.n_states = 4096. That's why I was confused whats wrong with this model. Maybe @zakari-denis can help me.

*   One-hot encoding looks like an interesting add-on to solve this problem... However, I don't know how to integrate it with this library. May I request some assistance or further references to proceed with it?

Thank you all in advance for your help!!
